File: src/main/java/at/searles/parsing/combinators/ReducerJoin.java
package at.searles.parsing.combinators;

import at.searles.parsing.*;
import at.searles.parsing.printing.PartialConcreteSyntaxTree;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ReducerJoin<T> implements Reducer<T, T> {

    private final Reducer<T, T> parserReducer;
    private final Reducer<T, T> printerReducer;

    public ReducerJoin(Recognizer separator, Reducer<T, T> reducer) {
        this.parserReducer = Reducer.opt(reducer.then(Reducer.rep(separator.then(reducer))));
        this.printerReducer = Reducer.opt(Reducer.rep(reducer.then(separator)).then(reducer));
    }

    @Nullable
    @Override
    public T parse(ParserStream stream, @NotNull T left) {
        return parserReducer.parse(stream, left);
    }

    @Nullable
    @Override
    public PartialConcreteSyntaxTree<T> print(@NotNull T t) {
        return printerReducer.print(t);
    }

    @Override
    public boolean recognize(ParserStream stream) {
        return parserReducer.recognize(stream);
    }

    @Override
    public String toString() {
        return String.format("join(%s)", parserReducer.toString());
    }
}

File: src/main/java/at/searles/parsing/combinators/ReducerJoin.kt
package at.searles.parsing.combinators

import at.searles.parsing.ParserStream
import at.searles.parsing.Recognizer
import at.searles.parsing.Reducer
import at.searles.parsing.Reducer.Companion.opt
import at.searles.parsing.Reducer.Companion.rep
import at.searles.parsing.printing.PartialTree

class ReducerJoin<T>(private val reducer: Reducer<T, T>, private val separator: Recognizer) : Reducer<T, T> {

    private val parserReducer: Reducer<T, T> = (reducer + (separator + reducer).rep()).opt()
    private val printerReducer: Reducer<T, T> = ((reducer + separator).rep() + reducer).opt()

    override fun parse(left: T, stream: ParserStream): T? {
        return parserReducer.parse(left, stream)
    }

    override fun print(item: T): PartialTree<T>? {
        return printerReducer.print(item)
    }

    override fun recognize(stream: ParserStream): Boolean {
        return parserReducer.recognize(stream)
    }

    override fun toString(): String {
        return "$reducer.rep($separator)"
    }

}